### PR TITLE
Intersection with terms

### PR DESCRIPTION
**Pull request by elshize**

### CODE REVIEW

**Comment by elshize** [comment]
@amallia Can you take a look at this? I get all scores equal 0 (I even printed out the entire `results` and all are in fact zeroes). Am I doing something wrong?

**Comment by amallia** [comment]
> @amallia Can you take a look at this? I get all scores equal 0 (I even printed out the entire `results` and all are in fact zeroes). Am I doing something wrong?

@elshize I have tested it and I am able to retrieve the score of the intersection. Steps to replicate my experiment:

### Create the index with the test data:
```./bin/create_freq_index -t pefopt -c ../test/test_data/test_collection  -o test.pefopt```
### Create the additional metadata:
```./bin/create_wand_data -c ../test/test_data/test_collection -o test.wand```

### Perform the intersection:
```./bin/compute_intersection -t pefopt -i test.pefopt -w test.wand -q ../test/test_data/queries```

### The following is my output:
```
0	12	10.7526
1	8	10.7121
2	7	10.1845
3	0	0
4	284	0.817719
5	19	10.7458
6	27	7.59847
7	0	0
8	1	15.7422
9	20	7.88953
10	21	8.93683
11	0	0
12	103	7.44573
13	0	0
14	1	6.25612
15	6	6.78048
16	15	12.3746
17	3	8.3576
18	1	8.86082
19	8	8.17585
20	0	0
21	325	0.942618
22	1134	3.5261
23	8	9.00911
24	130	9.7026
25	1	7.16875
26	58	11.6152
27	3	9.2513
28	0	0
29	0	0
30	1	6.64212
31	3	8.34363
32	117	8.82377
33	1	5.63655
34	0	0
35	50	0.644731
36	0	0
37	9	0.629966
38	0	0
39	0	0
40	377	8.58282
41	6	9.87135
42	13	13.1381
43	0	0
44	0	0
45	0	0
46	0	0
47	0	0
48	0	0
49	0	0
50	3	12.7899
51	19	9.99783
52	21	10.5353
53	3	27.7018
54	8	10.8135
55	11	10.9942
56	0	0
57	9	0.482994
58	0	0
59	19	11.1702
60	0	0
61	3	9.44242
62	0	0
63	0	0
64	7	10.5645
65	0	0
66	0	0
67	1	16.5755
68	9	12.7193
69	5	12.4509
70	10	14.2997
71	6	14.3731
72	4	5.57711
73	2	7.18717
74	0	0
75	127	0.904065
76	4	7.9726
77	1	3.79551
78	1	9.37333
79	0	0
80	4	10.4002
81	0	0
82	36	8.23128
83	1	7.58331
84	0	0
85	0	0
86	10	0.729877
87	0	0
88	0	0
89	3	10.9083
90	0	0
91	1	8.80528
92	0	0
93	0	0
94	0	0
95	47	11.4697
96	3	10.3956
97	6	13.0063
98	3	0.847606
99	22	12.2965
100	25	11.6912
101	1	12.2824
102	0	0
103	2	12.6857
104	12	13.3596
105	356	0.856192
106	0	0
107	5	20.7461
108	9	12.5931
109	0	0
110	0	0
111	72	12.4888
112	0	0
113	1	2.94668
114	0	0
115	71	9.20318
116	0	0
117	1	9.62311
118	0	0
119	1	7.02756
120	0	0
121	0	0
122	1	15.6416
123	0	0
124	0	0
125	0	0
126	0	0
127	0	0
128	2	10.3397
129	87	8.01282
130	1	13.9492
131	0	0
132	0	0
133	10	9.05595
134	0	0
135	0	0
136	4	13.336
137	37	7.84569
138	1	7.19674
139	17	9.62623
140	0	0
141	0	0
142	1	5.78999
143	108	8.95713
144	0	0
145	6	19.6142
146	0	0
147	0	0
148	58	8.37668
149	1	5.9837
150	0	0
151	34	10.0586
152	1	6.96685
153	0	0
154	0	0
155	45	11.9548
156	5	13.4858
157	0	0
158	3	11.7044
159	12	13.3251
160	12	12.7256
161	5	10.5206
162	0	0
163	1	7.20786
164	0	0
165	5	7.9187
166	15	11.4383
167	0	0
168	0	0
169	3	14.3193
170	6	14.5203
171	0	0
172	0	0
173	1067	0.985914
174	0	0
175	0	0
176	11	8.72142
177	0	0
178	51	13.9454
179	33	12.7824
180	123	0.911901
181	12	10.083
182	2	8.66683
183	0	0
184	0	0
185	1	5.65323
186	0	0
187	0	0
188	1	6.83248
189	42	11.1242
190	0	0
191	0	0
192	0	0
193	4	0.623653
194	271	6.27217
195	5	12.0846
196	0	0
197	0	0
198	96	0.820038
199	1	11.2032
200	0	0
201	0	0
202	0	0
203	0	0
204	1	11.4891
205	2	7.10457
206	0	0
207	0	0
208	0	0
209	0	0
210	0	0
211	0	0
212	0	0
213	0	0
214	16	10.8162
215	61	10.7756
216	22	0.988097
217	2	8.58733
218	0	0
219	161	7.15711
220	64	0.898705
221	0	0
222	4	9.07846
223	0	0
224	0	0
225	17	9.44447
226	0	0
227	0	0
228	0	0
229	0	0
230	70	8.14426
231	1	8.92662
232	22	10.5238
233	2	11.2134
234	14	10.9787
235	2	9.68445
236	4	13.8435
237	954	1.87112
238	21	9.49725
239	0	0
240	820	0.969986
241	15	12.4913
242	0	0
243	16	11.3455
244	11	10.2164
245	0	0
246	0	0
247	8	7.1866
248	11	13.3743
249	1	0.670063
250	5	9.20825
251	5	14.2338
252	11	0.621761
253	0	0
254	2	7.64676
255	807	0.983891
256	170	0.950008
257	109	10.1199
258	7	10.2098
259	1	13.6576
260	0	0
261	24	8.70158
262	1	7.68502
263	31	9.15594
264	0	0
265	82	9.29002
266	1	0.486916
267	1	11.562
268	928	3.7514
269	0	0
270	0	0
271	13	0.857063
272	0	0
273	16	9.26764
274	36	10.8537
275	1	13.0495
276	1808	0.984441
277	0	0
278	436	17.2084
279	49	10.4135
280	393	0.954056
281	98	7.66571
282	235	6.04751
283	0	0
284	0	0
285	0	0
286	0	0
287	0	0
288	93	8.77368
289	0	0
290	0	0
291	38	9.79235
292	0	0
293	0	0
294	0	0
295	1094	3.26971
296	49	9.78577
297	1	7.70101
298	1	14.3354
299	0	0
300	0	0
301	8	11.8016
302	8	8.06361
303	1	4.8417
304	0	0
305	0	0
306	0	0
307	0	0
308	3	0.630855
309	1	16.2116
310	16	9.1226
311	76	10.0168
312	1	12.0254
313	0	0
314	4	9.05984
315	2	0.0960458
316	4	13.6844
317	4	9.78057
318	0	0
319	0	0
320	2	11.9966
321	1	9.02319
322	0	0
323	1	8.90822
324	347	7.75614
325	111	9.07175
326	3	17.1798
327	7	10.261
328	0	0
329	1	0.148425
330	0	0
331	0	0
332	89	11.6591
333	0	0
334	46	10.0216
335	22	9.9946
336	0	0
337	0	0
338	25	12.8281
339	12	9.17711
340	1	6.48163
341	0	0
342	13	11.7686
343	0	0
344	0	0
345	0	0
346	3	8.50473
347	13	12.953
348	135	5.66061
349	32	0.867954
350	177	11.0329
351	0	0
352	572	5.14228
353	15	0.869643
354	0	0
355	66	9.7705
356	167	8.00795
357	0	0
358	15	0.650515
359	0	0
360	7	19.2467
361	15	11.4648
362	0	0
363	0	0
364	124	0.956067
365	0	0
366	18	9.51428
367	0	0
368	2	11.7526
369	1	0.620114
370	28	10.6861
371	8	12.2466
372	3	8.47669
373	0	0
374	0	0
375	0	0
376	0	0
377	42	0.9456
378	1	9.87189
379	307	0.925483
380	0	0
381	19	9.61451
382	201	5.34545
383	0	0
384	2	7.20446
385	43	11.027
386	13	12.4519
387	2	18.906
388	2	13.5662
389	0	0
390	22	11.023
391	0	0
392	0	0
393	1	10.7325
394	0	0
395	0	0
396	34	0.958319
397	10	9.81761
398	8	9.39232
399	0	0
400	23	11.8175
401	0	0
402	6	14.1895
403	8	0.597797
404	25	7.78057
405	0	0
406	15	0.730761
407	13	13.5882
408	158	7.24506
409	16	8.90102
410	5	8.97873
411	0	0
412	0	0
413	0	0
414	6	0.452127
415	0	0
416	7	12.2347
417	321	8.0068
418	0	0
419	222	6.79713
420	0	0
421	13	8.4043
422	0	0
423	5	12.3566
424	3	9.83581
425	1	25.0252
426	6	15.0554
427	0	0
428	17	12.0889
429	1	0.262261
430	1	10.527
431	4	9.55881
432	5	9.81356
433	0	0
434	0	0
435	0	0
436	0	0
437	1	4.05108
438	0	0
439	2	5.4742
440	0	0
441	60	11.6448
442	0	0
443	0	0
444	3	7.26233
445	1	10.6605
446	1	9.37904
447	0	0
448	0	0
449	3	8.29084
450	172	0.967596
451	0	0
452	19	9.62056
453	20	12.1869
454	0	0
455	1	0.617854
456	0	0
457	0	0
458	0	0
459	4	12.2033
460	25	11.3945
461	1	8.3686
462	0	0
463	0	0
464	4	16.4321
465	2	13.2815
466	0	0
467	0	0
468	0	0
469	0	0
470	8	10.6175
471	4	11.9245
472	1	0.47074
473	1	5.20432
474	14	9.90362
475	1	4.39092
476	1	0.145306
477	0	0
478	9	11.6467
479	280	0.921502
480	0	0
481	4	15.864
482	51	13.268
483	28	10.6239
484	1	3.61213
485	78	9.06614
486	0	0
487	45	10.1734
488	76	12.4381
489	4	8.80831
490	2	8.0046
491	2	8.90722
492	25	0.831466
493	1	4.50969
494	359	6.79077
495	7	7.97252
496	5	7.49868
497	1	6.63128
498	5	12.3409
499	4	0.585134
```

**Comment by elshize** [comment]
Thanks, I guess there must be something wrong with my index. I'll test again.

**Comment by elshize** [comment]
I found that the single lists don't return the correct max score. @amallia can you see what I did wrong there?

**Review by amallia — APPROVED** [review]

File: include/pisa/intersection.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    enum class IntersectionType { Query, Combinations };
+
+    /// Mask is set to a static-sized bitset for performance.
+    /// Using dynamic bitset slows down a lot, and these operations will be performed on
+    /// much shorter queries than the limit allows for. It makes very little sense to do
+    /// that on anything longer.
+    constexpr std::size_t MAX_QUERY_LEN_EXP = 31;
+    constexpr std::size_t MAX_QUERY_LEN = 1 << MAX_QUERY_LEN_EXP;
+    using Mask = std::bitset<MAX_QUERY_LEN_EXP>;
```
Comment (amallia) [inline include/pisa/intersection.hpp:22]: Kinda overcomplicated, but ok!

**Pull request merged**